### PR TITLE
PBM-896: physical restore of encrypted DBs

### PR DESCRIPTION
**Pull request by dAdAbird**
**Pass "security" config options from the host cluster to tmp mongod processes.**

**Introduce partlyDone status for physical restore.**
It will help in cases where nodes in the replica set have different keys.
So we can restore at least one node in each replica set.in the replica 
set have different keys. So we can restore at least one node in each replica set.

Since we restore physical bcps on all nodes in the replset, we may treat
a success on at least one node in each replset as a partial success.
Meaning the cluster can be started with the restore although some nodes
gonna be in initiallSync.

For that, if at least one node in the replset moved
to the "done", the replset is "partlyDone". And a replset is "done" if all
nodes moved to "done". For cluster success, all replsets must move either
to "done" or "partlyDone". Cluster
is "partlyDone" if at least one replset
is "partlyDone".

Also `pbm describe-restore` shows the restore status accordingly.

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Logical | Physical |
| ------------- | ------------- | ------------- |
| Sharded psmdb 4.2 | failed | passed |
| Sharded psmdb 4.4 | passed | passed |
| Sharded psmdb 5.0 | passed | passed |
| New cluster psmdb 4.2 | passed | - |
| New cluster psmdb 4.4 | passed | - |
| New cluster psmdb 5.0 | passed | - |
| Non-sharded psmdb 4.2 | passed | failed |
| Non-sharded psmdb 4.4 | passed | failed |
| Non-sharded psmdb 5.0 | passed | failed |
| Single-node psmdb 4.2 | failed | failed |
| Single-node psmdb 4.4 | failed | failed |
| Single-node psmdb 5.0 | failed | failed |

**Comment by JNKPercona** [comment]
| Test name  | Logical | Physical |
| ------------- | ------------- | ------------- |
| Sharded psmdb 4.2 | passed | passed |
| Sharded psmdb 4.4 | passed | passed |
| Sharded psmdb 5.0 | failed | passed |
| New cluster psmdb 4.2 | passed | - |
| New cluster psmdb 5.0 | passed | - |
| Non-sharded psmdb 4.2 | failed | passed |
| New cluster psmdb 4.4 | passed | - |
| Non-sharded psmdb 4.4 | failed | passed |
| Single-node psmdb 4.4 | passed | passed |
| Non-sharded psmdb 5.0 | passed | passed |
| Single-node psmdb 4.2 | passed | passed |
| Single-node psmdb 5.0 | failed | failed |

File: pbm/bsontypes.go [hunk 1]
```diff
@@ -237,7 +237,29 @@ type MongodOpts struct {
...
+		TokenFile  *string `bson:"tokenFile,omitempty" json:"tokenFile,omitempty" yaml:"tokenFile,omitempty"`
+		Secret     *string `bson:"secret,omitempty" json:"secret,omitempty" yaml:"secret,omitempty"`
+	} `bson:"vault,omitempty" json:"vault,omitempty" yaml:"vault,omitempty"`
+	KMIP *struct {
+		ServerName            *string `bson:"serverName,omitempty" json:"serverName,omitempty" yaml:"serverName,omitempty"`
+		Port                  *int    `bson:"port,omitempty" json:"port,omitempty" yaml:"port,omitempty"`
+		ClientCertificateFile *string `bson:"clientCertificateFile,omitempty" json:"clientCertificateFile,omitempty" yaml:"clientCertificateFile,omitempty"`
+		ClientKeyFile         *string `bson:"clientKeyFile,omitempty" json:"clientKeyFile,omitempty" yaml:"clientKeyFile,omitempty"`
+		ServerCAFile          *string `bson:"serverCAFile,omitempty" json:"serverCAFile,omitempty" yaml:"serverCAFile,omitempty"`
+		KeyIdentifier         *string `bson:"keyIdentifier,omitempty" json:"keyIdentifier,omitempty" yaml:"keyIdentifier,omitempty"`
```
Comment (olexandr-havryliak) [inline pbm/bsontypes.go:261]: Please, also add **kmipClientCertificatePassword** 
https://github.com/percona/percona-server-mongodb/blob/v5.0/src/mongo/db/mongod_options_encryption.idl#L151-L158

**Comment by JNKPercona** [comment]
| Test name  | Logical | Physical |
| ------------- | ------------- | ------------- |
| Sharded psmdb 4.2 | failed | passed |
| Sharded psmdb 4.4 | failed | passed |
| Sharded psmdb 5.0 | failed | failed |
| New cluster psmdb 4.2 | passed | - |
| New cluster psmdb 5.0 | failed | - |
| New cluster psmdb 4.4 | passed | - |
| Non-sharded psmdb 4.2 | failed | passed |
| Non-sharded psmdb 4.4 | passed | passed |
| Single-node psmdb 4.2 | passed | failed |
| Non-sharded psmdb 5.0 | passed | passed |
| Single-node psmdb 4.4 | passed | failed |
| Single-node psmdb 5.0 | passed | failed |

File: pbm/bsontypes.go (hunk 1 above)
Comment (dAdAbird) [reply pbm/bsontypes.go]: https://github.com/percona/percona-backup-mongodb/pull/697/commits/b08da272d7636b161a13f26bc36e281f78c8773e

**Review by sandraromanchenko — CHANGES_REQUESTED** [review]

File: pbm/bsontypes.go [hunk 2]
```diff
@@ -237,7 +237,30 @@ type MongodOpts struct {
...
+	Security *MongodOptsSec    `bson:"security,omitempty" json:"security,omitempty" yaml:"security,omitempty"`
+}
+
+type MongodOptsSec struct {
+	EnableEncryption     *bool   `bson:"enableEncryption,omitempty" json:"enableEncryption,omitempty" yaml:"enableEncryption,omitempty"`
+	EncryptionCipherMode *string `bson:"encryptionCipherMode,omitempty" json:"encryptionCipherMode,omitempty" yaml:"encryptionCipherMode,omitempty"`
+	EncryptionKeyFile    *string `bson:"encryptionKeyFile,omitempty" json:"encryptionKeyFile,omitempty" yaml:"encryptionKeyFile,omitempty"`
+	RelaxPermChecks      *bool   `bson:"relaxPermChecks,omitempty" json:"relaxPermChecks,omitempty" yaml:"relaxPermChecks,omitempty"`
+	Vault                *struct {
+		ServerName *string `bson:"serverName,omitempty" json:"serverName,omitempty" yaml:"serverName,omitempty"`
```
[the remark this answers is not in the document]
Comment (dAdAbird) [reply pbm/bsontypes.go]: https://github.com/percona/percona-backup-mongodb/pull/697/commits/cbf813badcdb3ff83d52c2b863c28fb1aea2583d

**Comment by JNKPercona** [comment]
| Test name  | Logical | Physical |
| ------------- | ------------- | ------------- |
| Sharded psmdb 4.2 | failed | failed |
| Sharded psmdb 5.0 | failed | failed |
| New cluster psmdb 4.2 | passed | - |
| New cluster psmdb 5.0 | passed | - |
| New cluster psmdb 4.4 | passed | - |
| Non-sharded psmdb 4.2 | failed | passed |
| Non-sharded psmdb 4.4 | failed | passed |
| Sharded psmdb 4.4 | - | failed |
| Single-node psmdb 4.2 | failed | failed |
| Single-node psmdb 5.0 | failed | failed |
| Single-node psmdb 4.4 | passed | failed |
| Non-sharded psmdb 5.0 | failed | passed |

**Comment by JNKPercona** [comment]
| Test name  | Logical | Physical |
| ------------- | ------------- | ------------- |
| New cluster psmdb 4.2 | passed | - |
| New cluster psmdb 4.4 | passed | - |
| New cluster psmdb 5.0 | passed | - |
| Non-sharded psmdb 4.4 | failed | passed |
| Non-sharded psmdb 4.2 | failed | passed |
| Non-sharded psmdb 5.0 | failed | passed |
| Single-node psmdb 4.2 | failed | passed |
| Single-node psmdb 4.4 | failed | passed |
| Single-node psmdb 5.0 | failed | passed |
| Sharded psmdb 4.2 | failed | passed |
| Sharded psmdb 4.4 | failed | failed |
| Sharded psmdb 5.0 | passed | failed |

**Review by sandraromanchenko — APPROVED** [review]

**Pull request merged**